Paypal Express Hold Authorization failure still allows order to process

swilkinson-bluebolt opened the pull request:
Overview:
This change fixes the issue we found for the Paypal Express whereby an order will process from a Hold despite the fact that the user's primary and backup funding source had insufficient funds.

Issue as reported:
Our client received an order from Hotcakes to there ERP that had a failed Paypal Hold authorization.  The checkout process should not allowed an order to go through with a failed authorization.  Therefore they had to contact the customer for payment.

How to recreate:
1) In Hotcakes admin, enable Paypal Express payment option and select the "Hold" option for the fund type
2) Create a Paypal sandbox user that has a valid, verified account with a backup credit card funding source.  Ensure the user account does not have sufficient funds for a purchase.  In addition, ensure the test user's backup funding source does not have sufficient funds.
3) Checkout using the Paypal Express method.  Notice the order will complete and in the Hotcakes Order admin, the Payment transactions will show the failed Hold authorization.

This Pull Request Fix:
My fix will inspect the response of the Paypal authorize call and if it was unsuccessful, it will include that as part of setting the OrderTransaction.Success flag rather than just ignoring it.  That will return the user back to checkout page to try the payment again.

Thanks!
Scott

hismightiness commented:
Changed the target to the `development` branch.

hismightiness commented on Website/Portals/_default/HotcakesViews/_default/Views/Cart/MiniCart.cshtml:
```diff
@@ -1,74 +1,18 @@
         </a>
     </div>
-    @if (!Model.CartEmpty)
+    @if (Model.TotalQuantity > 0)
```
This doesn't appear to be a good update for stores that allow zero dollar orders.

hismightiness commented on Website/web.config:
```diff
@@ -28,6 +28,7 @@
     </sectionGroup>
     <sectionGroup name="system.web.webPages.razor" type="System.Web.WebPages.Razor.Configuration.RazorWebSectionGroup, System.Web.WebPages.Razor">
       <section name="pages" type="System.Web.WebPages.Razor.Configuration.RazorPagesSection, System.Web.WebPages.Razor" requirePermission="false" />
+      <section name="host" type="System.Web.WebPages.Razor.Configuration.HostSection, System.Web.WebPages.Razor" requirePermission="false" />
```
Shouldn't this update be in the viewset web.config and not the DNN website config?

hismightiness commented:
One more thing, @swilkinson-bluebolt ... You may want to review this PR by Matt Rutledge since you both did a SPA mini cart.
https://github.com/HotcakesCommerce/hotcakes-commerce-core/pull/98

swilkinson-bluebolt commented on .gitignore:
```diff
@@ -43,7 +43,6 @@
 *.dotCover
 *.ldf
 MSBuildLog.txt
-_[Bb]uild/
```
Not sure about that, disregard it

swilkinson-bluebolt commented on Libraries/Hotcakes.Commerce.Dnn/Web/HotcakesModuleBase.cs:
```diff
@@ -182,10 +182,10 @@ protected virtual void InitializeScriptManager()
                 }
 
-                var scriptResource = string.Format(" {1} var hcc = hcc || {{}}; hcc.l10n = {0};", json, analyticsScript);
+                var scriptResource = string.Format(" {1}\n var hcc = hcc || {{}};\n hcc.l10n = {0};\n", json, analyticsScript);
```
It fixes nothing but allows you to debug the lines in the browser's inspector

mtrutledge commented:
Mine was done so long ago I bet @swilkinson-bluebolt's is better.

On Wed, Oct 9, 2019 at 1:31 PM Will Strohl <notifications@github.com> wrote:

> One more thing, @swilkinson-bluebolt
> <https://github.com/swilkinson-bluebolt> ... You may want to review this
> PR by Matt Rutledge since you both did a SPA mini cart.
> #98 <https://github.com/HotcakesCommerce/hotcakes-commerce-core/pull/98>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/HotcakesCommerce/hotcakes-commerce-core/pull/212?email_source=notifications&email_token=AAQIT7XWFFSLOOGWOZ4WTC3QNYWPRA5CNFSM4IRWIUEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAZBTIQ#issuecomment-540154274>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQIT7X5CYUBII4OG2XC3CDQNYWPRANCNFSM4IRWIUEA>
> .
>

swilkinson-bluebolt commented on Libraries/Hotcakes.Commerce.Dnn/Web/HotcakesModuleBase.cs:
```diff
@@ -182,10 +182,10 @@ protected virtual void InitializeScriptManager()
                 if (HccApp.CurrentStore.Settings.Analytics.UseGoogleEcommerce)
                 {
-                    analyticsScript = " ga('require', 'ecommerce'); ";
+                    analyticsScript = "try { ga('require', 'ecommerce'); }\n catch(err) {}";
```
Somehow this PR got combined with other fixes.  Sorry about that.  I will look at separating them.

swilkinson-bluebolt commented on Website/DesktopModules/Hotcakes/Core/Admin/Orders/Default.aspx:
```diff
@@ -68,6 +68,9 @@
                     <asp:LinkButton ID="btnGo" runat="server" Text="Filter Results" CssClass="hcIconRight" OnClick="btnGo_Click" />
                 </div>
+                <div class="hcFormItem hcUnderCheck">
+                    <asp:CheckBox ID="chkFilterIsOrderNum" runat="server" Checked="True" Text="Search criteria is an Order # (fast lookup)" AutoPostBack="False" />
```
This was supposed to be in a separate PR.  I didn't localize because I saw other text hardcoded.  But I can add to a resource file.

hismightiness replied:
That would be great.  Please do use the resource file.  Thanks!  :)

hismightiness commented:
I like the latest commit.  Thanks, @swilkinson-bluebolt .  Do you know when the other items in the review might be addressed?  :)

hismightiness commented:
@swilkinson-bluebolt Would you be open to doing a screenshare meeting to go over all of this?  There are a lot of updates this in this one PR and they should have been in separate PRs.  Also, there are questions I and that I'm unsure if they were addressed.

hismightiness commented:
@swilkinson-bluebolt PING! :D

hismightiness commented:
@swilkinson-bluebolt Double-PING!  :D

hismightiness commented:
Hey @swilkinson-bluebolt ... Another commit?  Let's get this PR done together, finally!  :D

swilkinson-bluebolt commented:
Will,

There is a problem with the pull request and branching.  If you have time, let’s talk next week about how we can clean this up.  If anything, this was a small change.  I could probably just take you through the changes and you could commit it.  Let me know if you have any availability.

Thanks!
Scott

From: Will Strohl <notifications@github.com>
Sent: Tuesday, December 1, 2020 10:09 PM
To: HotcakesCommerce/hotcakes-commerce-core <hotcakes-commerce-core@noreply.github.com>
Cc: Scott Wilkinson <swilkinson@blueboltsolutions.com>; Mention <mention@noreply.github.com>
Subject: Re: [HotcakesCommerce/hotcakes-commerce-core] Paypal Express Hold Authorization failure still allows order to process (#212)


Hey @swilkinson-bluebolt<https://github.com/swilkinson-bluebolt> ... Another commit? Let's get this PR done together, finally! :D

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/HotcakesCommerce/hotcakes-commerce-core/pull/212#issuecomment-736977040>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKPN5VSYJZ5EBF27VVSCSIDSSW4XPANCNFSM4IRWIUEA>.

hismightiness commented:
Finally....! He lives!  :D 
Yes.  Let's do that, please.

swilkinson-bluebolt commented:
Your Mightiness,

I do live 😊.  Forgive my delays on this.

I deleted the old pull request from our master fork.  That was causing all kinds of confusion.

I rebased my fork repo with your development branch, made the paypal fix again, and submitted a clean pull request.  Hopefully that will take care of it.  If you have any questions, just send them to my Bluebolt email address.

Thanks!

Scott Wilkinson
Solutions Architect
224.558.8649
[cid:image001.png@01D6CE2A.8E9F9660]



From: Will Strohl <notifications@github.com>
Sent: Tuesday, December 8, 2020 6:52 PM
To: HotcakesCommerce/hotcakes-commerce-core <hotcakes-commerce-core@noreply.github.com>
Cc: Scott Wilkinson <swilkinson@blueboltsolutions.com>; Mention <mention@noreply.github.com>
Subject: Re: [HotcakesCommerce/hotcakes-commerce-core] Paypal Express Hold Authorization failure still allows order to process (#212)


Finally....! He lives! :D
Yes. Let's do that, please.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/HotcakesCommerce/hotcakes-commerce-core/pull/212#issuecomment-741346030>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKPN5VT46IHU2LRH7D2VR4DST3C3TANCNFSM4IRWIUEA>.
